### PR TITLE
fix(pages): render custom errors for notFound results

### PR DESCRIPTION
**Pull request by NathanDrake2406**
## Overview

| Area | Details |
| --- | --- |
| Goal | Match Next.js when a Pages Router data method returns `notFound` and the HTML request must render the configured 404/error page. |
| Core change | `resolvePagesPageData()` now returns an HTML `notFound` signal instead of constructing the built-in 404 response directly. |
| Main boundary | The generated Pages entry owns rerouting to `pages/404` or `pages/_error`, while data requests still return the JSON-shaped 404 envelope. |
| Primary files | `pages-page-data.ts`, `pages-server-entry.ts`, `pages-get-initial-props.ts`, `dev-server.ts`, `tests/pages-router.test.ts`. |
| Expected impact | `_error.getInitialProps(ctx)` receives the original request URL and `asPath` for `getStaticProps` or `getServerSideProps` `notFound` results. |

## Why

A Pages Router `notFound` result is not the final HTML response. In Next.js, `getStaticProps` marks render metadata as not found, the Pages route handler delegates to `render404()`, and the server then resolves `pages/404` before falling back to `pages/_error`. That fallback render still builds the normal `NextPageContext`, including `ctx.req`, `ctx.res`, and `ctx.asPath`.

| Area | Principle / invariant | What this PR changes |
| --- | --- | --- |
| HTML `notFound` | Data resolution should describe the route outcome, not preempt the configured error-page renderer. | HTML `notFound` returns `{ kind: "notFound" }`; only `/_next/data` requests keep returning `{}` with status 404. |
| Error route rendering | `pages/404` and `pages/_error` should be resolved by the Pages entry with the original visible URL. | The generated entry rerenders the configured not-found route with `statusCode: 404` and the original `asPath`. |
| `getInitialProps` context | Pages `getInitialProps` belongs at the page render boundary and should observe sent-response short-circuits. | A typed helper invokes page `getInitialProps` in prod and dev without widening production code through `any` or broad assertions. |

## What changed

| Scenario | Before | After |
| --- | --- | --- |
| `getStaticProps` returns `{ notFound: true }` for an HTML request | vinext returned the built-in default 404, skipping custom `pages/_error`. | vinext rerenders `pages/404` or `pages/_error` and preserves 404 status. |
| `_error.getInitialProps(ctx)` on that rerender | Not called, so `ctx.req?.url` and `ctx.asPath` never reached the page. | Called with `req.url` and `asPath` from the original request, matching the upstream fixture. |
| `/_next/data/...json` not found | Returned JSON-shaped 404. | Preserved. |

<details>
<summary>Maintainer review path</summary>

1. `packages/vinext/src/server/pages-page-data.ts` - the resolver now emits a typed `notFound` signal for HTML while preserving JSON data-request behavior.
2. `packages/vinext/src/entries/pages-server-entry.ts` - the generated Pages runtime reroutes that signal to explicit `/404` or `/_error`, avoiding catch-all matches and recursion.
3. `packages/vinext/src/server/pages-get-initial-props.ts` - shared typed loader for Pages component `getInitialProps`, including the sent-response exception from Next.js `loadGetInitialProps`.
4. `packages/vinext/src/server/dev-server.ts` - dev parity for page and error-page `getInitialProps` context.
5. `tests/pages-router.test.ts` and `tests/pages-page-data.test.ts` - upstream fixture port plus lower-boundary contract coverage.

</details>

<details>
<summary>Validation</summary>

- `vp check`
- `vp test run tests/pages-page-data.test.ts`
- `vp test run tests/pages-router.test.ts`
- Commit hook: `vp test run tests/entry-templates.test.ts`
- Commit hook: checked formatting, lint, types, and `knip`
- Upstream deploy suite: `vp env exec --node 24 ./scripts/run-nextjs-deploy-suite.sh /Users/nathan/Projects/vinext/.refs/nextjs-v16.2.6 --retries 0 -c 1 --debug test/e2e/error-handler-not-found-req-url/error-handler-not-found-req-url.test.ts`

</details>

<details>
<summary>Risk / compatibility</summary>

- Public API: no new public API. The new helper is internal.
- Runtime: affects Pages Router HTML `notFound` handling and Pages component `getInitialProps` invocation when no gSP/gSSP owns props.
- Data requests: preserved as JSON 404 envelopes so client-router data fetches can still parse the response.
- Compatibility: intentionally follows Next.js routing order for `notFound` fallback to `pages/404` and `pages/_error`.
- Existing-app risk: apps that relied on vinext's built-in 404 shortcut for gSP/gSSP `notFound` will now see their configured error page, which is the Next.js-compatible behavior.

</details>

## References

| Reference | Why it matters |
| --- | --- |
| [Next.js upstream test: error-handler-not-found-req-url](https://github.com/vercel/next.js/blob/ee6e79b1792a/test/e2e/error-handler-not-found-req-url/error-handler-not-found-req-url.test.ts) | The fixture this PR ports and passes through the vinext deploy harness. |
| [Next.js `getStaticProps` `notFound` metadata](https://github.com/vercel/next.js/blob/ee6e79b1792a/packages/next/src/server/render.tsx#L873-L880) | Shows `notFound` becoming render metadata instead of an immediate built-in response. |
| [Next.js render returns null for `metadata.isNotFound`](https://github.com/vercel/next.js/blob/ee6e79b1792a/packages/next/src/server/render.tsx#L971-L975) | Explains why the Pages handler owns the final 404 render decision. |
| [Next.js Pages handler delegates `isNotFound` to `render404()`](https://github.com/vercel/next.js/blob/ee6e79b1792a/packages/next/src/server/route-modules/pages/pages-handler.ts#L585-L591) | The production behavior vinext now mirrors. |
| [Next.js `render404` and error fallback](https://github.com/vercel/next.js/blob/ee6e79b1792a/packages/next/src/server/base-server.ts#L2574-L2658) | Shows the `/404` then `/_error` fallback sequence. |
| [Next.js context passed to `loadGetInitialProps`](https://github.com/vercel/next.js/blob/ee6e79b1792a/packages/next/src/server/render.tsx#L746-L797) | Confirms `ctx.req`, `ctx.res`, and `ctx.asPath` are part of the error render context. |
| [Next.js `loadGetInitialProps`](https://github.com/vercel/next.js/blob/ee6e79b1792a/packages/next/src/shared/lib/utils.ts#L345-L393) | The sent-response and invalid-return semantics used by the vinext helper. |
| [Next.js docs for `getStaticProps.notFound`](https://github.com/vercel/next.js/blob/ee6e79b1792a/docs/02-pages/04-api-reference/03-functions/get-static-props.mdx#L106-L108) | Documents that `notFound` should render a 404 page. |

### CODE REVIEW

**Comment by NathanDrake2406** [comment]
Thanks for the review! Addressing the three observations:

**1. `_app.getInitialProps` not invoked (pre-existing gap) — no change in this PR**

Confirmed not introduced by this PR. Searching for `loadGetInitialProps` and App gIP usage in `packages/vinext/src/server/`:

- `dev-server.ts` and `prod-server.ts` only call `loadPagesGetInitialProps` on the page component, never on `App`.
- The generated `pages-server-entry.ts` `_renderPage` invokes `loadPagesGetInitialProps(options.pageModule.default, ...)` (the page module), not the App wrapper.
- `_app.tsx` is wrapped around the page element in the React tree, but vinext doesn't extract a `getInitialProps` static from it.

This is the pre-existing behavior — the page gIP path works, the App gIP path is a known gap unrelated to notFound handling. I agree a tracking issue is worth filing (will do separately) but fixing it here would expand the PR's scope significantly (requires App gIP loading + `{ pageProps }` wrapping in both dev/prod). Out of scope for this fix.

**2. Stricter-than-Next.js return validation (deliberate divergence) — no change**

The `isPropsObject` rejection of arrays/functions/primitives in `pages-get-initial-props.ts:79-87` is intentional defense-in-depth. Next.js's `if (!props)` check accepts `[]` (a truthy array) which then crashes later when downstream code does `Object.assign(pageProps, initialProps)` or `JSON.stringify` over the page data. The stricter check throws early with a clear error message that names the offending component and value. Keeping this as a deliberate vinext improvement.

**3. Response-sent check asymmetry — fixed**

Implemented in `a595ea8`. Exported `isResponseSent` from `pages-get-initial-props.ts` and replaced both `if (res.headersSent)` checks in `pages-page-data.ts` (gSSP at line 487, gIP at line 697) with `if (isResponseSent(res))`. The prod `PagesReqResResponse` type only declares `headersSent` (its getter is backed by the `ended` flag in `pages-node-compat.ts:203-204`), so the prod-only inline check was functionally correct — but using the helper future-proofs the prod resolver against response-shape changes and unifies sent-detection with the dev path and the helper itself.

Added tests:
- `short-circuits getServerSideProps responses when only writableEnded is set`
- `short-circuits getInitialProps responses when only writableEnded is set`
- Focused unit test for the exported `isResponseSent` covering both prod-shaped (`{ headersSent }`) and node-shaped (`{ writableEnded }`) mocks, plus the undefined/null edge cases.

**Comment by NathanDrake2406** [comment]
Addressed the three observations from the latest review:

**1. `createGsspReqRes()` allocation gating — fixed in `81badc98`.** Exported `hasPagesGetInitialProps` from `pages-get-initial-props.ts` and gated the prod block in `resolvePagesPageData` on it. The hot path for plain static pages (no gSSP, no gSP, no gIP) is now allocation-free — `createGsspReqRes()` is only called when there is actually a `getInitialProps` function present.

**2. Sent-response normalizes non-objects to `{}` — no change.** This is a deliberate, documented divergence from Next.js behavior. The caller spreads the result into `pageProps`, so a non-object return would crash — normalizing to `{}` avoids that crash and gives a clearer error path.

**3. Missing dev-only instance-method guard — no change.** Minor dev-ergonomics improvement beyond the scope of this PR.

File: packages/vinext/src/server/pages-page-data.ts [hunk 1]
```diff
@@ -676,6 +677,35 @@ export async function resolvePagesPageData(
     }
   }
 
+  if (
+    typeof options.pageModule.getServerSideProps !== "function" &&
+    typeof options.pageModule.getStaticProps !== "function"
+  ) {
+    const { req, res, responsePromise } = options.createGsspReqRes();
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/server/pages-page-data.ts:689]: Addressed in 81badc98. Exported `hasPagesGetInitialProps` from `pages-get-initial-props.ts` and gated the prod block on it. The hot path for plain static pages (no gSSP, no gSP, no gIP) is now allocation-free.
